### PR TITLE
cosma: added package at v2.0.2

### PR DESCRIPTION
**Pull request by teonnik**
Distributed Communication-Optimal Matrix-Matrix Multiplication Library

https://github.com/eth-cscs/COSMA

### CODE REVIEW

**Comment by alalazo** [comment]
Can you fix the `flake8` issues reported by Travis?

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    Distributed Communication-Optimal Matrix-Matrix Multiplication Library
+    """
+
+    homepage = "https://github.com/eth-cscs/COSMA"
+    maintainers = ['teonnik', 'kabicm']
+
+    # note: The default archives produced with github do not have the archives
+    #       of the submodules.
+    version('develop',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]:
```suggestion
    version('master',
```

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            git='https://github.com/eth-cscs/COSMA.git',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: Move this to the top of the file below homepage

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            branch='master',
+            submodules=True)
+    version('2.0.2',
+            url='https://github.com/eth-cscs/COSMA/releases/download/v2.0.2/cosma.tar.gz',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: Move this to the top of the file below homepage

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    # version('1.0.0',
+    #         url='https://github.com/eth-cscs/COSMA/releases/download/1.0/cosma.tar.gz',
+    #         sha256='c142104258dcca4c17fa7faffc2990a08d2777235c7980006e93c5dca51061f6')
+
+    variant('blas', default='mkl',
+            values=('mkl', 'openblas', 'netlib', 'cuda', 'rocm'),
+            description='BLAS backend')
+    variant('scalapack', default='none',
+            values=('mkl', 'netlib'),
+            description='Optional ScaLAPACK support.')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: The way Spack handles these is through virtual dependencies. Basically, you add a dependency on `blas` or `scalapack`, and then you set your default BLAS/LAPACK/SCALAPACK provider in `packages.yaml` or on the command-line. So you can remove these variants.

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        if 'blas=mkl' in spec:
+            args += ['-DCOSMA_BLAS=MKL']
+        elif 'blas=netlib' in spec:
+            args += ['-DCOSMA_BLAS=CUSTOM']
+        elif 'blas=openblas' in spec:
+            args += ['-DCOSMA_BLAS=OPENBLAS']
+        elif 'blas=cuda' in spec:
+            args += ['-DCOSMA_BLAS=CUDA']
+        else:  # 'blas=rocm' in spec:
+            args += ['-DCOSMA_BLAS=ROCM']
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: You can check which BLAS provider was chosen by either:
```python
if spec['blas'].name == 'intel-mkl':
```
or:
```python
if '^mkl' in spec:
```

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    version('2.0.2',
+            url='https://github.com/eth-cscs/COSMA/releases/download/v2.0.2/cosma.tar.gz',
+            sha256='4f3354828bc718f3eef2f0098c3bdca3499297497a220da32db1acd57920c68d')
+    # note: this version fails to build at the moment
+    # version('1.0.0',
+    #         url='https://github.com/eth-cscs/COSMA/releases/download/1.0/cosma.tar.gz',
+    #         sha256='c142104258dcca4c17fa7faffc2990a08d2777235c7980006e93c5dca51061f6')
+
+    variant('cuda', default=False)
+    variant('scalapack', default=False)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: These variants need descriptions.

**Comment by teonnik** [comment]
Since ScaLAPACK and CUDA are optional dependencies, I've added variants for them. I am not sure if there is a cleaner way to express that.

```
variant('cuda', default=False)
variant('scalapack', default=False)
```

**Comment by adamjstewart** [comment]
Can you move the `git` and `url` attributes to the package-level so new versions don't need to duplicate them?

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            env.set('CUDA_PATH', self.spec['cuda'].prefix)
+
+    def cmake_args(self):
+        spec = self.spec
+        args = ['-DCOSMA_WITH_TESTS=OFF',
+                '-DCOSMA_WITH_APPS=OFF',
+                '-DCOSMA_WITH_PROFILING=OFF',
+                '-DCOSMA_WITH_BENCHMARKS=OFF']
+
+        if '^intel-mkl' in spec:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: I would change this to `^mkl`, same below. Both intel-mkl and intel-parallel-studio+mkl provide the mkl virtual dependency

Comment (teonnik) [reply var/spack/repos/builtin/packages/cosma/package.py]: I am not convinced `+mkl` is needed for intel-parallel-studio: https://github.com/spack/spack/pull/15567 . Would it be OK to rewrite this to `'^intel-mlk' or '^intel-parallel-studio'` instead? MKL is always shipped with intel-parallel-studio.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cosma/package.py]: You changed the one below but not this one:
```suggestion
        if '^mkl' in spec:
```

**Pull request merged**